akka-projection-cassandra/src/main/scala/akka/projection/cassandra/internal/CassandraProjectionImpl.scala

```diff
@@ -226,3 +159,151 @@ import akka.stream.scaladsl.Source
+
+  /*
+   * Build the final AtLeastOnceSettings to use, if currently set to None fallback to values in config file
+   */
+  private[akka] def atLeastOnceSettingsOrDefaults(
+      implicit systemProvider: ClassicActorSystemProvider): AtLeastOnceSettings =
+    atLeastOnceSettings.getOrElse(AtLeastOnceSettings(systemProvider))
+
+  protected[projection] def composedSource(
+      offsetStore: CassandraOffsetStore,
+      source: Source[(Offset, Envelope), NotUsed],
+      handlerFlow: Flow[(Offset, Envelope), Offset, NotUsed])(
+      implicit systemProvider: ClassicActorSystemProvider): Source[Done, NotUsed] = {
+    val atLeastOnceSettings = atLeastOnceSettingsOrDefaults
+
+    import atLeastOnceSettings._
+
+    (saveOffsetAfterEnvelopes, saveOffsetAfterDuration) match {
+      case (1, _) =>
+        // optimization of general AtLeastOnce case
+        source.via(handlerFlow).mapAsync(1) { offset =>
+          offsetStore.saveOffset(projectionId, offset)
+        }
+      case (afterEnvelopes, orAfterDuration) =>
+        source
+          .via(handlerFlow)
+          .groupedWithin(afterEnvelopes, orAfterDuration)
+          .collect { case grouped if grouped.nonEmpty => grouped.last }
+          .mapAsync(parallelism = 1) { offset =>
+            offsetStore.saveOffset(projectionId, offset)
+          }
+    }
+  }
+
+  override def withSettings(settings: ProjectionSettings): AtLeastOnceCassandraProjectionImpl[Offset, Envelope] =
+    new AtLeastOnceCassandraProjectionImpl(projectionId, sourceProvider, Option(settings), atLeastOnceSettings, handler)
+
+  override def withAtLeastOnceSettings(
+      settings: AtLeastOnceSettings): AtLeastOnceCassandraProjectionImpl[Offset, Envelope] =
+    new AtLeastOnceCassandraProjectionImpl(projectionId, sourceProvider, projectionSettings, Some(settings), handler)
+
+  override def withSaveOffsetAfterEnvelopes(afterEnvelopes: Int): AtLeastOnceCassandraProjectionImpl[Offset, Envelope] =
+    new AtLeastOnceCassandraProjectionImpl(
+      projectionId,
+      sourceProvider,
+      projectionSettings,
+      atLeastOnceSettings.map(_.withSaveOffsetAfterEnvelopes(afterEnvelopes)),
+      handler)
+
+  override def withSaveOffsetAfterDuration(
+      afterDuration: FiniteDuration): AtLeastOnceCassandraProjectionImpl[Offset, Envelope] =
+    new AtLeastOnceCassandraProjectionImpl(
+      projectionId,
+      sourceProvider,
+      projectionSettings,
+      atLeastOnceSettings.map(_.withSaveOffsetAfterDuration(afterDuration)).getOrElse(AtLeastOnceSettings.),
+      handler)
+
+  /**
+   * Java API
+   */
+  override def withSaveOffsetAfterDuration(
+      afterDuration: java.time.Duration): AtLeastOnceCassandraProjectionImpl[Offset, Envelope] =
+    new AtLeastOnceCassandraProjectionImpl(
+      projectionId,
+      sourceProvider,
+      projectionSettings,
+      atLeastOnceSettings.map(_.withSaveOffsetAfterDuration(afterDuration)),
+      handler)
+}
```

Comment:
These implementations don't work correctly (causing some tests to fail) because `atLeastOnceSettings` is always `None` by default. In order to load defaults we need an actor system to load defaults from config, but we only provide a system provider when we _run_ the projection.  To fix this we would either have to 

1. Provide a system provider for each `with*` call
2. Provide a system provider when we first create the projection in the projection factory methods